### PR TITLE
Prevent loading both `json/ext` and `json/pure`

### PR DESCRIPTION
**Pull request by casperisfine**
Fix: https://github.com/ruby/json/issues/650
Ref: https://github.com/ruby/json/issues/646

`json_pure` currently doesn't work well at all because `json/ext` is a default gem, so if you have code requiring `json/pure` and some other code requiring `json`, you end up with both loaded.

If the `json` and `json_pure` versions match, it's not too bad, but if they don't match, you might run into issues with private APIs no longer matching.

### CODE REVIEW

File: lib/json/ext.rb [hunk 1]
```diff
@@ -2,24 +2,26 @@
...
-    else
-      require 'json/ext/parser'
-      require 'json/ext/generator'
-      $DEBUG and warn "Using Ext extension for JSON."
-      JSON.parser = Parser
-      JSON.generator = Generator
-    end
+unless defined?(::JSON::JSON_LOADED)
+  # Prevent json/pure from being loaded at the same time
+  $LOADED_FEATURES << File.expand_path("../pure.rb", __FILE__)
```
Comment (casperisfine) [inline lib/json/ext.rb:7]: I tired to add some stub `json/ext/generator.rb` etc, but it was making loading the test suite very contrived.

In the end I think marking the alternative implementation as loaded is OK?

File: lib/json_pure.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
+
+require 'json/pure'
```
Comment (casperisfine) [inline lib/json_pure.rb:3]: This is a bit orthogonal, but ensure that adding `gem "json_pure"` in a Gemfile properly load `json/pure` in the expected order.

File: json_pure.gemspec [hunk 3]
```diff
@@ -16,36 +16,27 @@ Gem::Specification.new do |s|
...
+    "CHANGES.md",
+    "COPYING",
+    "BSDL",
+    "LEGAL",
+    "README.md",
+    "json.gemspec",
+    *(Dir["lib/**/*.rb"] - Dir["lib/json/ext/**/*.rb"]),
+  ]
+
   s.files = [
```
Comment (eregon) [inline json_pure.gemspec:30]: You're setting the files twice here

Comment (casperisfine) [reply json_pure.gemspec]: Good catch. Fixed.

File: json_pure.gemspec [hunk 4]
```diff
@@ -16,36 +16,27 @@ Gem::Specification.new do |s|
...
   s.rdoc_options = ["--title", "JSON implementation for ruby", "--main", "README.md"]
+
+  s.files = [
+    "CHANGES.md",
+    "COPYING",
+    "BSDL",
+    "LEGAL",
+    "README.md",
+    "json.gemspec",
+    *(Dir["lib/**/*.rb"] - Dir["lib/json/ext/**/*.rb"]),
```
[the remark this answers is not in the document]
Comment (casperisfine) [reply json_pure.gemspec:27]: >  as you could get a mix of which files get loaded from where, especially if different versions.

That's the opposite. This way if you got `json_pure` above `json` in your gemfile, it takes precedence.

The goal is exactly for both gems to define these "features" (as in `$LOADED_FEATURES`) so we prevent mixed loading.

> IMO if one's environment is too broken to install extension gems, then they simply can't install it, and I don't think json needs to workaround such broken environments.

Some people don't want compilers in production, but yes, IMO the value of `json_pure` is really doubtful now that `json` is a default gem, if you can't compile extensions, just figure out how to use the stdlib.

But in the meantime I think this PR is a decent way forward.

File: json_pure.gemspec [hunk 5]
```diff
@@ -16,36 +16,17 @@ Gem::Specification.new do |s|
...
-    "lib/json/add/symbol.rb",
-    "lib/json/add/time.rb",
-    "lib/json/common.rb",
-    "lib/json/ext.rb",
-    "lib/json/generic_object.rb",
-    "lib/json/pure.rb",
-    "lib/json/pure/generator.rb",
-    "lib/json/pure/parser.rb",
-    "lib/json/version.rb",
+    "json.gemspec",
```
Comment (eregon) [inline json_pure.gemspec:26]:
```suggestion
    "json_pure.gemspec",
```
or not including it at all

File: lib/json/ext.rb (hunk 1 above)
Comment (eregon) [reply lib/json/ext.rb:7]: Given that file doesn't exist (because of the changes in json.gemspec), does it have any effect?
It feels brittle.
Is it necessary? I would think the `JSON::JSON_LOADED` check is enough.

**Comment by eregon** [comment]
Also in that scenario `lib/json/common.rb` would be loaded from json_pure, but maybe there is some code there which is not compatible with what the C extension expects.

File: lib/json/ext.rb (hunk 1 above)
Comment (casperisfine) [reply lib/json/ext.rb:7]: It's in case the version of the "other gem" is much older.

**Comment by casperisfine** [comment]
Hum, right, I think I need to use more `require_relative`.

**Comment by eregon** [comment]
I think we should just have `json_pure` an empty gem depending on `json`, it seems intrinsically very brittle to have 2 gems which share some Ruby files (e.g. `lib/json/common.rb` and `lib/json/version.rb`) of potentially different versions.

If we wanted to just make sure the versions match we'd need to have json_pure depend on json same version, which is the same thing and then I think there is no point to ship the same files in two gems.
